### PR TITLE
Propagate CONTAINERS_CONF to conmon

### PR DESCRIPTION
**Pull request by dgibson**
The `CONTAINERS_CONF` environment variable can be used to override the configuration file, which is useful for testing.  However, at the moment this variable is not propagated to conmon.  That means in particular, that conmon can't propagate it back to podman when invoking `--exit-command`.  The mismatch in configuration between the starting and cleaning up podman instances can cause a variety of errors.

This patch also adds two related test cases.  One checks explicitly that the correct `$CONTAINERS_CONF` value appears in conmon's environment.  The other checks for a possible specific impact of this bug: if we use a nonstandard name for the runtime (even if its path is just a regular crun), then the podman container cleanup invoked at container exit will fail. That has the effect of meaning that a container started with `-d --rm` won't be correctly removed once complete.

### CODE REVIEW

**Comment by rhatdan** [comment]
Great catch @dgibson 
LGTM
/approve
@containers/podman-maintainers PTAL

File: libpod/oci_conmon_linux.go
Comment (Luap99) [inline libpod/oci_conmon_linux.go:1324]:
```suggestion
	conf, ok := os.LookupEnv("CONTAINERS_CONF")
	if ok {
		env = append(env, fmt.Sprintf("CONTAINERS_CONF=%s", conf))
	}
```
We should also support an empty containers conf env var.

**Review by Luap99 — COMMENTED** [review]
one nit otherwise LGTM

**Comment by Luap99** [comment]
@edsantiago PTAL at tests

File: test/system/800-config.bats
Comment (edsantiago) [inline test/system/800-config.bats:8]: Broken, missing end quote, this will fail. But please stay tuned for further review feedback.

File: test/system/800-config.bats
Comment (edsantiago) [inline test/system/800-config.bats:29]: Suggestion: `run_podman rm -f -t 0 "$cid"`

File: test/system/800-config.bats
Comment (edsantiago) [inline test/system/800-config.bats:51]: You've run the container with `--rm`. This is a NOP.

File: test/system/800-config.bats
Comment (edsantiago) [inline test/system/800-config.bats:46]: You don't actually seem to be testing anything here. Friendly suggestion:
```bash
    run_podman run --name foo $IMAGE true
    run_podman container inspect -l --format '{{.OCIRuntime}}'
    is "$output" "nonstandard_runtime_name" "[engine]->runtime override"

    run_podman rm foo
```

File: test/system/800-config.bats
Comment (edsantiago) [inline test/system/800-config.bats:15]: There is no guarantee that `$OCI_RUNTIME` will be defined.

File: test/system/800-config.bats
Comment (edsantiago) [inline test/system/800-config.bats:12]: Please don't do this. Instead, do:
```
    conf_tmp="$PODMAN_TMPDIR/containers.conf"
    cat >$conf_tmp ...

   ...

   CONTAINERS_CONF=$conf_tmp run_podman .....
```
Reason: `export`ing will not propagate across tests, but it _will_ carry over into `teardown()`. If for any reason the resulting conf file is bad, the `podman` invocations in `teardown` will fail, and cascading test failures will ensue.

**Review by edsantiago — CHANGES_REQUESTED** [review]
Thanks for the PR! Good catch, but tests need a few fixes.

File: libpod/oci_conmon_linux.go
Comment (dgibson) [reply libpod/oci_conmon_linux.go]: Makes sense, I've folded that in.

File: test/system/800-config.bats
Comment (dgibson) [reply test/system/800-config.bats:8]: Weird... I'm not sure how I missed that when I ran the tests.  Anyway, fixed now.

File: test/system/800-config.bats
Comment (dgibson) [reply test/system/800-config.bats:29]: Done, thanks.

File: test/system/800-config.bats
Comment (dgibson) [reply test/system/800-config.bats:51]: It *should* be a no-op, but precisely the bug I'm addressing here means it isn't without this fix.

File: test/system/800-config.bats
Comment (dgibson) [reply test/system/800-config.bats:46]: I'm testing that the `podman rm` *fails* with a no such container error.

File: test/system/800-config.bats
Comment (dgibson) [reply test/system/800-config.bats:12]: Updated as suggested.

File: test/system/800-config.bats
Comment (dgibson) [reply test/system/800-config.bats:15]: Good point.  I've changed it to be `${OCI_RUNTIME:-crun}` instead.

File: test/system/800-config.bats
Comment (dgibson) [reply test/system/800-config.bats:46]: I've put a comment to explain this admittedly counter-intuitive logic.  I hope that will address both this and your previous comment.

File: test/system/800-config.bats
Comment (edsantiago) [inline test/system/800-config.bats:15]: Just to save you the time of waiting for CI to finish: this is likely to fail on one of our CI hosts, in which only `runc` is installed. You may want to get the default from `podman info` instead.

File: test/system/800-config.bats
Comment (edsantiago) [inline test/system/800-config.bats:50]: Swoon. I love comments like these. Thank you!

File: libpod/oci_conmon_linux.go
Comment (edsantiago) [inline libpod/oci_conmon_linux.go:1322]: I haven't checked (it's almost bedtime here), but what if `CONTAINERS_CONF=` (i.e., is defined but empty)?

**Review by edsantiago — COMMENTED** [review]
Thank you for your attention to all these details. Some comments, one suggestion, and I'm off for the evening. I'll check again in the (my) morning and hope to see green CI.

File: test/system/800-config.bats
Comment (edsantiago) [inline test/system/800-config.bats:28]: Oh, one more: you can remove the comment (no longer applicable) and the `CONTAINERS_CONF=` (not needed for `rm`)

**Comment by TomSweeneyRedHat** [comment]
LGTM
once @edsantiago 's concerns are addressed.

File: test/system/800-config.bats
Comment (dgibson) [reply test/system/800-config.bats]: Good idea, implemented.

File: test/system/800-config.bats
Comment (dgibson) [reply test/system/800-config.bats:32]: Well.. it's applicable in the sense that that's the reason we have `-f -t 0` to force an immediate SIGKILL.

We'd probably be fine without the `CONTAINERS_CONF` setting in this case, but the container was created with the alternate configuration, so AFAICT it makes sense to use the same configuration when removing it.

File: libpod/oci_conmon_linux.go
Comment (dgibson) [reply libpod/oci_conmon_linux.go:1322]: I changed this from `GetEnv` to `LookupEnv` at @Luap99's suggestion to handle exactly that case.

File: test/system/800-config.bats
Comment (dgibson) [reply test/system/800-config.bats:32]: I also note that the `-t` option doesn't appear to be documented in `man podman-rm` :/

Comment (edsantiago) [reply test/system/800-config.bats:32]: >force an immediate SIGKILL.

Oh. Yeah, I forgot that that's weird (I just consider it annoying). You can't interrupt `sleep`. (Wish I could get that patch for myself). OK, leaving the `CONTAINERS_CONF=` is fine, I can see your point.

Re `-t`: it's relatively new, might not have propagated to your install yet. Rest assured we have hooks that guarantee that all non-hidden options are documented in man pages.

**Comment by edsantiago** [comment]
Failure looks real:
```
         # # podman-remote [...] rm SHA
         # bbb7cc96e3077f7e7508649e6355dd838fd9bff4e787921d6bd7c78e1292994b
         # #/vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv
         # #| FAIL: exit code is 0; expected 1
         # #\^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Should this test be `skip_if_remote`-ed?

**Comment by dgibson** [comment]
> Failure looks real:
> 
> ```
>          # # podman-remote [...] rm SHA
>          # bbb7cc96e3077f7e7508649e6355dd838fd9bff4e787921d6bd7c78e1292994b
>          # #/vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv
>          # #| FAIL: exit code is 0; expected 1
>          # #\^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> ```
> 
> Should this test be `skip_if_remote`-ed?

I don't think so, although I'm not sure, since I know almost nothing about how podman remote works.

AFAICT failing in the way it appears to means that the `--rm` option really is failing to work, which seems like a real bug, and I doubt it's caused by my code change, merely exposed by the new test.  I don't really know why my change which fixes the local case would not fix the remote case - but, again, I really don't know how remote works.  I don't think I'll be able to debug this on my own.

**Comment by edsantiago** [comment]
It's a race condition:
```console
$ while :;do cid=$(bin/podman-remote run -d --rm quay.io/libpod/testimage:20210610 true);bin/podman-remote rm $cid;done
f67da6c90c667f7f53a4f35c88b501661424646b47e3c21ee9f9d876d9854750
8ddffb3e971f6986fef5fd21a57c98fe28bdaa70eb66371cbcf6b69395cc1993
Error: no container with name or ID "ee8a170a78727763a88ccb08319f04899ffe94b7de1a76bea2a33d7d6c47a439" found: no such container
Error: no container with name or ID "021aa341a56f9dfab681307eefc1a832226190277897b1e3b4b40319f76e0531" found: no such container
1efc9f1d56946353c51fc7b420f5de274fca7af5af2cfec495edc33300aba629
82f3cf3d2310b85cb2424701a98bcf956981957d93d76dae23a06556259eedda
825797a3f88cdeabcc55be736aecba25c0cab743044e36260705136cbfca9d01
Error: no container with name or ID "d8cdcc884b0013a1fa87d68fac7de274d16f4bc66c8b8ad14bfc3ae975c538f6" found: no such container
30ca114f4ffab6ac87e3741c37251524bdc207fc1a7a9e13e31cbf56a88a9108
a9e78031c688df5868b0197f01ef3c94f6372b00f786bfeb1236ad32acf792e1
d3709db21ac0e1044dc9748f83b1e03ebc8a60b367024a0746e387d4ea102077
```
...and it's an understandable one, in your test, because there really can't be any timing guarantees here. This race affects both local and remote, it just seems probable that remote is slightly slower.

Proposed solution: I don't know. Maybe `run_podman wait $cid`? I would recommend checking out a broken version of podman (one that demonstrates the bug you're trying to test), then figuring out how to reliably give the `run -d --rm` process time to complete and clean up (or not, in the case of the bug). Then write your test to conform to that.

BTW please also remove the `export` from the second test.

**Comment by dgibson** [comment]
> Oh - and `CONTAINERS_CONF` doesn't work over remote anyway (I think), so you probably do want to `skip_if_remote` just like in your first test.

Ok, I've done that for now.  Do you know how we could find out definitively if `$CONTAINERS_CONF` is supposed to work over remote?

> This does not relieve you from fixing the race condition, though: it's a real one, and could conceivably trigger on a real run. It's convenient that podman-remote reproduces it so easily, because that can help you write a robust wait test.

Given I have no idea how to configure or invoke remote podman, that doesn't actually help that much.  But, I can force that race to happen easily enough using a `sleep 3` instead of a `true`.  I think I have something solid now, lets see what the CI thinks.

**Comment by dgibson** [comment]
Alas, the CI is now failing on the e2e ginkgo testsuite, for reasons I can't discern :(

**Comment by edsantiago** [comment]
The [plugin in containers failure](https://storage.googleapis.com/cirrus-ci-6707778565701632-fcae48/artifacts/containers/podman/5933325356367872/html/int-podman-fedora-35-root-container.log.html#t--use-plugin-in-containers--1) looks very real to me: either an unintended consequence of your changes, or a bug in the test. Basically, that test code sets CONTAINERS_CONF, and is now failing. I don't work Fridays so I can't look into it further today; but I thought you needed to know.

The sys remote failure looks like a flake, but it's not one that looks familiar to me, and has already been restarted once, and there's no point in restarting it since there's a genuine failure anyway.

File: test/system/800-config.bats
Comment (edsantiago) [inline test/system/800-config.bats:63]: Suggestion:
```bash
   run_podman '?' wait "$cid"
   if [[ $status != 0 ]]; then
      is "$output" "Error:.*no container" "unexpected error from podman wait"
   fi
```

**Comment by rhatdan** [comment]
@dgibson  Is this something we need to get into podman 4.0, or can it wait for podman 4.1?

**Comment by dgibson** [comment]
> @dgibson Is this something we need to get into podman 4.0, or can it wait for podman 4.1?

I have no strong opinion.  This is definitely a bug, but it tends only to bite in some unusual edge cases.

File: test/system/800-config.bats
Comment (dgibson) [reply test/system/800-config.bats:63]: Oh, good idea.  I'd forgotten about the `?` option in the `run_podman` helper.

**Comment by dgibson** [comment]
At least some of the CI failures I'm getting are consistent, however, I'm finding myself in deeply unfamiliar weeds amongst volume plugins trying to debug them.  I *think* the bug I'm addressing here was just masking a pre-existing bug (in the code or the tests, I'm not sure).

There are several failures, but focussing on one to start with.  It occurs in this section of test code from `tests/e2e/volume_plugin_test.go`:

```
// HACK: `volume rm -f` is timing out trying to remove containers using the volume.
// Solution: remove them manually...
// TODO: fix this when I get back
rmAll := podmanTest.Podman([]string{"rm", "-af"})
rmAll.WaitWithDefaultTimeout()
Expect(rmAll).Should(Exit(0))
```

Specifically, the `podman rm -af` is failing with exit code 125 and the message `Error: cannot remove volume testVolume1 from plugin testvol4, but it has been removed from Podman: required plugin missing` (this does seem to be exposed anywhere obvious - I deduced this with a lot of stracing).  I have no idea why propoagating `CONTAINERS_CONF` further would cause this error.  I thought it might be that the bug I'm addressing caused the need for the HACK in the first place, but that doesn't seem to be the case - if I remove this stanza it fails during cleanup as described by the comment.

/cc @mheon since he added this test.

**Comment by mheon** [comment]
I think I know what's going on; the container holding the volume plugin is being `rm -f`'d before one of the containers using the plugin. Not sure how this hasn't been flaking before...

**Comment by mheon** [comment]
I think https://github.com/containers/podman/pull/13264 should resolve this once it merges

**Comment by dgibson** [comment]
> I think I know what's going on; the container holding the volume plugin is being `rm -f`'d before one of the containers using the plugin. Not sure how this hasn't been flaking before...

Aha!  One of the possible consequences of this bug is that automatic container cleanup can fail (because the cleanup process doesn't get the right configuration.  I wonder if that could be what was happening - the cleanup of the plugin container failed, meaning the later removal didn't fail.

**Comment by dgibson** [comment]
Thanks @mheon, local tests are looking good with your fix, next we'll see how the CI takes it.

**Comment by dgibson** [comment]
@edsantiago, @mheon has fixed the CI failures, and I believe I've addressed all the changes you requested.  Can you re-review, so we can merge this?

**Comment by mheon** [comment]
LGTM

**Review by edsantiago — APPROVED** [review]

**Comment by edsantiago** [comment]
/lgtm

Thanks for following up.

**Pull request merged**
